File: web/src/app/infrastructure/rest/ride.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from '../../../env/environment';
import { CreateRideRequest } from './model/create-ride.model';
import { LocationDto } from './model/location.model';
import { EndRideRequest, RideCancellationRequest, RideResponse } from './model/ride-response.model';

export interface PageResponse<T> {
  content: T[];
  totalElements?: number;
  totalPages?: number;
  number?: number;
  size?: number;
}

export interface RoutePoint {
  location: LocationDto;
  order: number;
}

export interface RideEstimationResponse {
  estimatedTimeInMinutes: number;
  estimatedCost: number;
  estimatedDriverArrivalInMinutes: number;
  estimatedDistance: number;
  routePoints: RoutePoint[];
}

export interface AdminStatsResponse {
  activeRidesCount: number;
  averageDriverRating: number;
  driversOnlineCount: number;
  totalPassengersInRides: number;
}

@Injectable({
  providedIn: 'root'
})
export class RideService {
  private apiUrl = `${environment.apiHost}rides`;

  constructor(private http: HttpClient) {}

  estimateRide(request: CreateRideRequest): Observable<RideEstimationResponse> {
    return this.http.post<RideEstimationResponse>(`${this.apiUrl}/estimate`, request);
  }

  getRidesHistory(params: {
    page: number;
    size: number;
    sort?: string;
    fromDate?: string; 
    toDate?: string;
    driverId?: number;
    passengerId?: number;
    status?: string;
  }): Observable<PageResponse<RideResponse>> {
    let queryParams: any = {
      page: params.page.toString(),
      size: params.size.toString(),
    };

    if (params.sort) queryParams.sort = params.sort;
    if (params.fromDate) queryParams.fromDate = params.fromDate;
    if (params.toDate) queryParams.toDate = params.toDate;
    if (params.driverId) queryParams.driverId = params.driverId.toString();
    if (params.passengerId) queryParams.passengerId = params.passengerId.toString();
    if (params.status && params.status !== 'all') queryParams.status = params.status;

    return this.http.get<PageResponse<RideResponse>>(this.apiUrl, { params: queryParams });
  }

  orderRide(request: CreateRideRequest): Observable<RideResponse> {
    return this.http.post<RideResponse>(`${this.apiUrl}`, request);
  }

  getRide(id: number): Observable<RideResponse> {
    return this.http.get<RideResponse>(`${this.apiUrl}/${id}`);
  }

  getActiveRide(userId?: number): Observable<RideResponse | null> {
    const query = userId ? `?userId=${encodeURIComponent(String(userId))}` : '';
    return this.http.get<RideResponse>(`${this.apiUrl}/active${query}`).pipe(
      catchError(() => of(null))
    );
  }

  startRide(id: number): Observable<RideResponse> {
    return this.http.put<RideResponse>(`${this.apiUrl}/${id}/start`, {});
  }

  endRide(id: number, request: EndRideRequest): Observable<RideResponse> {
    return this.http.put<RideResponse>(`${this.apiUrl}/${id}/end`, request);
  }

  cancelRide(id: number, request: RideCancellationRequest): Observable<RideResponse> {
    return this.http.put<RideResponse>(`${this.apiUrl}/${id}/cancel`, request);
  }

  addRouteToFavorites(rideId: number, favoriteRoute: any): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${rideId}/favourite-route`, favoriteRoute);
  }

  getFavoriteRoutes(passengerId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/${passengerId}/favourite-routes`);
  }

  removeFavoriteRoute(passengerId: number, routeId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${passengerId}/favourite-routes/${routeId}`);
  }

  stopRide(id: number, request: any): Observable<RideResponse> {
    return this.http.put<RideResponse>(`${this.apiUrl}/${id}/stop`, request);
  }

  /**
   * Get the active ride for a user (IN_PROGRESS first, then earliest PENDING/SCHEDULED)
   */
  getActiveRideForUser(userId: number): Observable<RideResponse | null> {
    return this.http.get<RideResponse>(`${this.apiUrl}/active?userId=${userId}`);
  }

  /**
   * Complete a stop during an in-progress ride
   */
  completeStop(rideId: number, stopIndex: number): Observable<RideResponse> {
    return this.http.put<RideResponse>(`${this.apiUrl}/${rideId}/stop/${stopIndex}/complete`, {});
  }

  /**
   * Report an inconsistency during a ride (passenger only)
   */
  reportInconsistency(rideId: number, request: { remark: string }): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${rideId}/inconsistencies`, request);
  }

  /**
   * Cancel ride as passenger (reason is optional)
   */
  cancelRideAsPassenger(id: number, reason?: string): Observable<RideResponse> {
    return this.http.put<RideResponse>(`${this.apiUrl}/${id}/cancel`, { reason: reason || '' });
  }

  /**
   * Trigger panic button for a ride (driver or passenger)
   */
  panicRide(id: number, reason?: string): Observable<RideResponse> {
    return this.http.put<RideResponse>(`${this.apiUrl}/${id}/panic`, { reason: reason || null });
  }

  /**
   * Admin: Get all active rides (PENDING, ACCEPTED, SCHEDULED, IN_PROGRESS)
   */
  getAllActiveRides(params?: {
    page?: number;
    size?: number;
    sort?: string;
    search?: string;
    status?: string;
    vehicleType?: string;
  }): Observable<PageResponse<RideResponse>> {
    const query = new URLSearchParams();
    if (params?.page != null) query.set('page', String(params.page));
    if (params?.size != null) query.set('size', String(params.size));
    if (params?.sort) query.set('sort', params.sort);
    if (params?.search) query.set('search', params.search);
    if (params?.status) query.set('status', params.status);
    if (params?.vehicleType) query.set('vehicleType', params.vehicleType);

    const qs = query.toString();
    return this.http.get<PageResponse<RideResponse>>(`${this.apiUrl}/active/all${qs ? `?${qs}` : ''}`);
  }

  /**
   * Admin: Get dashboard statistics
   */
  getAdminStats(): Observable<AdminStatsResponse> {
    return this.http.get<AdminStatsResponse>(`${environment.apiHost}admin/stats`);
  }
}

File: mobile/app/src/main/java/com/example/mobile/services/RideService.java
package com.example.mobile.services;

import com.example.mobile.models.CreateRideRequest;
import com.example.mobile.models.EndRideRequest;
import com.example.mobile.models.FavoriteRouteRequest;
import com.example.mobile.models.FavoriteRouteResponse;
import com.example.mobile.models.InconsistencyRequest;
import com.example.mobile.models.PageResponse;
import com.example.mobile.models.RideCancellationRequest;
import com.example.mobile.models.RideEstimationResponse;
import com.example.mobile.models.RidePanicRequest;
import com.example.mobile.models.RideResponse;
import com.example.mobile.models.RideStopRequest;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface RideService {

    /**
     * Estimate ride fare before booking.
     * POST /api/rides/estimate
     */
    @POST("api/rides/estimate")
    Call<RideEstimationResponse> estimateRide(@Body CreateRideRequest request);

    /**
     * Create a new ride request.
     * POST /api/rides
     */
    @POST("api/rides")
    Call<RideResponse> createRide(
        @Body CreateRideRequest request,
        @Header("Authorization") String token
    );

    /**
     * Get ride by ID.
     * GET /api/rides/{id}
     */
    @GET("api/rides/{id}")
    Call<RideResponse> getRide(
        @Path("id") long id,
        @Header("Authorization") String token
    );

    /**
     * Get ride history with pagination and filters.
     * GET /api/rides
     */
    @GET("api/rides")
    Call<PageResponse<RideResponse>> getRidesHistory(
        @Query("driverId") Long driverId,
        @Query("passengerId") Long passengerId,
        @Query("status") String status,
        @Query("fromDate") String fromDate,
        @Query("toDate") String toDate,
        @Query("page") Integer page,
        @Query("size") Integer size,
        @Query("sort") String sort,
        @Header("Authorization") String token
    );

    /**
     * Cancel a ride.
     * PUT /api/rides/{id}/cancel
     */
    @PUT("api/rides/{id}/cancel")
    Call<RideResponse> cancelRide(
        @Path("id") long id,
        @Body RideCancellationRequest request,
        @Header("Authorization") String token
    );

    /**
     * Get active rides with pagination and filters.
     * GET /api/rides (filtered by status)
     */
    @GET("api/rides")
    Call<PageResponse<RideResponse>> getActiveRides(
        @Query("driverId") Long driverId,
        @Query("passengerId") Long passengerId,
        @Query("status") String status,
        @Query("page") Integer page,
        @Query("size") Integer size,
        @Header("Authorization") String token
    );

    /**
     * Trigger PANIC button for a ride (driver or passenger).
     * PUT /api/rides/{id}/panic
     */
    @PUT("api/rides/{id}/panic")
    Call<RideResponse> panicRide(
        @Path("id") long id,
        @Body RidePanicRequest request,
        @Header("Authorization") String token
    );

    /**
     * End (finish) a ride normally.
     * PUT /api/rides/{id}/end
     */
    @PUT("api/rides/{id}/end")
    Call<RideResponse> endRide(
        @Path("id") long id,
        @Body EndRideRequest request,
        @Header("Authorization") String token
    );

    /**
     * Stop a ride early at the current vehicle location (driver only).
     * PUT /api/rides/{id}/stop
     */
    @PUT("api/rides/{id}/stop")
    Call<RideResponse> stopRide(
        @Path("id") long id,
        @Body RideStopRequest request,
        @Header("Authorization") String token
    );

    /**
     * Mark an intermediate stop as completed.
     * PUT /api/rides/{id}/stop/{stopIndex}/complete
     */
    @PUT("api/rides/{id}/stop/{stopIndex}/complete")
    Call<RideResponse> completeStop(
        @Path("id") long rideId,
        @Path("stopIndex") int stopIndex,
        @Header("Authorization") String token
    );

    /**
     * Report an inconsistency for a ride.
     * POST /api/rides/{id}/inconsistencies
     */
    @POST("api/rides/{id}/inconsistencies")
    Call<Void> reportInconsistency(
        @Path("id") long rideId,
        @Body InconsistencyRequest request,
        @Header("Authorization") String token
    );

    /**
     * Get favourite routes for a passenger.
     * GET /api/rides/{id}/favourite-routes
     */
    @GET("api/rides/{id}/favourite-routes")
    Call<List<FavoriteRouteResponse>> getFavoriteRoutes(
        @Path("id") Long passengerId,
        @Header("Authorization") String token
    );

    /**
     * Add a favourite route for a passenger.
     * POST /api/rides/{id}/favourite-route
     */
    @POST("api/rides/{id}/favourite-route")
    Call<Void> addFavouriteRoute(
        @Path("id") Long passengerId,
        @Body FavoriteRouteRequest request,
        @Header("Authorization") String token
    );

    /**
     * Remove a favourite route for a passenger.
     * DELETE /api/rides/{passengerId}/favourite-routes/{favouriteRouteId}
     */
    @DELETE("api/rides/{passengerId}/favourite-routes/{favouriteRouteId}")
    Call<Void> removeFavouriteRoute(
        @Path("passengerId") Long passengerId,
        @Path("favouriteRouteId") Long favouriteRouteId,
        @Header("Authorization") String token
    );
}
